src/main/java/org/elasticsearch/indices/recovery/ShardRecoveryHandler.java

```diff
@@ -345,10 +350,49 @@ public void run() throws InterruptedException {
+                                EmptyTransportResponseHandler.INSTANCE_SAME).txGet();
+                    } catch (RemoteTransportException e) {
+                        IOException corruptIndexException;
+                        // we realized that after the index was copied and we wanted to finalize the recovery
+                        // the index was corrupted:
+                        //   - maybe due to a broken segments file on an empty index (transferred with no checksum)
+                        //   - maybe due to old segments without checksums or length only checks
+                        if ((corruptIndexException = ExceptionsHelper.unwrapCorruption(e)) != null) {
+                            try {
+                                final Store.MetadataSnapshot recoverySourceMetadata = store.getMetadata(snapshot);
+                                StoreFileMetaData[] metadata = Iterables.toArray(recoverySourceMetadata, StoreFileMetaData.class);
+                                ArrayUtil.timSort(metadata, new Comparator<StoreFileMetaData>() {
+                                    @Override
+                                    public int compare(StoreFileMetaData o1, StoreFileMetaData o2) {
+                                        return Long.compare(o1.length(), o2.length()); // check small flies first
+                                    }
+                                });
+                                for (StoreFileMetaData md : metadata) {
+                                    if (store.checkIntegrityNoException(md) == false) { // we are corrupted on the primary -- fail!
+                                        logger.warn("{} Corrupted file detected {} checksum mismatch", shard.shardId(), md);
+                                        // if we are not the first exception, add ourselves as suppressed to the main one:
+                                        throw e;
+                                    }
+                                }
+                            } catch (IOException ex) {
+                                e.addSuppressed(ex);
+                                throw e;
+                            }
+                            // corruption has happened on the way to replica
+                            RemoteTransportException exception = new RemoteTransportException("File corruption occurred on recovery but checksums are ok", null);
+                            exception.addSuppressed(e);
+                            exceptions.add(0, exception); // last exception first
+                            logger.warn("{} Remote file corruption during finalization on node {}, recovering {}. local checksum OK",
+                                    corruptIndexException, shard.shardId(), request.targetNode());
+
+                        } else {
+                            exceptions.add(0, e); // last exceptions first
+                        }
+                    }
```

Comment:
that is a leftover since I copied it from where we do thin in a different thread